src/directives/visibility.directive.ts

```diff
@@ -14,50 +15,26 @@ import {
  *
  */
 @Directive({ selector: '[visibilityObserver]' })
-export class VisibilityDirective implements OnInit, OnDestroy {
+export class VisibilityDirective {
 
   @HostBinding('class.visible') 
-  isVisible: boolean = false;
+  isVisible: boolean | undefined;
 
-  @Output() visible: EventEmitter<any> = new EventEmitter();
+  @Output() visible: EventEmitter<boolean> = new EventEmitter();
 
-  timeout: any;
+  constructor(private element: ElementRef) { }
 
-  constructor(private element: ElementRef, private zone: NgZone) { }
-
-  ngOnInit(): void {
-    this.runCheck();
-  }
-
-  ngOnDestroy(): void {
-    clearTimeout(this.timeout);
+  checkVisibility(): void {
+    const { offsetHeight, offsetWidth } = this.element.nativeElement;
+    this.onVisibilityChange(offsetHeight > 0 && offsetWidth > 0);
   }
 
-  onVisibilityChange(): void {
+  onVisibilityChange(visible: boolean): void {
+    if(visible === this.isVisible) return;
+    this.isVisible = visible;
     // trigger zone recalc for columns
-    this.zone.run(() => {
-      this.isVisible = true;
-      this.visible.emit(true);
+    setTimeout(() => {
```

Comment:
Why did you remove the zone run in favor of the settimeout?

Reply:
It was removed because it was already running inside of the angular zone. setTimeout was needed because of timing issues.

Reply:
I really hate doing `setTimeout` for timing issues. Do you know what specifically is the issue?

Reply:
I was looking into this pull request and to answer your question, without the setTimeout call I'm getting the following error at runtime.

Error: ExpressionChangedAfterItHasBeenCheckedError: Expression has changed after it was checked. Previous value: 'enabled: undefined'. Current value: 'enabled: true'.

I'm thinking that using setTimeout to cause the emit to be run outside of the current call stack is pushing the change of the value to after this error check.